README.md

```diff
@@ -93,75 +105,106 @@ internal GitHub server errors), the bot sleeps for 10 minutes
-* S-staging-checks-failed: Essentially duplicates GitHub "red x" mark for
-  the staging branch commit. The bot removes this label when it notices
+* `M-waiting-approvals`: The PR approval rules (see "Voting and PR
+  approvals" below) do not allow the bot to start merging regardless of
+  how old the PR gets. Compare with `M-waiting-objections`. The bot
+  removes this label when it notices that it is obsolete. TODO:
+  Implement.
+* `M-waiting-objections`: The PR approval rules (see "Voting and PR
+  approvals" below) allow the bot to start merging when the PR gets
+  older. The bot will eventually start merging this PR if no objections
+  are received while the bot is waiting. The wait time is determined by
+  the number of votes already received. The bot removes this label when
+  it notices that it is obsolete. TODO: Implement.
+* `M-waiting-staging-checks`: The bot determined that the PR is eligible
+  for merging, successfully completed the initial merging steps that are
+  supposed to trigger CI tests, and is now waiting for the CI test
+  results. The bot removes this label when it notices that all tests
+  have completed.
+* `M-passed-staging-checks`: Similar to the GitHub "green check" mark
+  for the staging branch commit (but ignores failures of optional
+  checks). Applied only if `config::skip_merge` option is on. The bot
+  removes this label when applying any other label (XXX: Except
+  M-failed-staging-checks. To improve code consistency and to avoid
+  lying in documentation, we should either remove this label only when
+  (re)setting the first label (i.e., M-waiting-staging-checks) or remove
+  this label when setting any other label, without making an exception
+  for M-failed-staging-checks. The implementation plans for the
+  currently unsupported labels may suggest the best way forward here.).
+* `M-failed-staging-checks`: Essentially duplicates GitHub "red x" mark
+  for the _staging commit_. The bot removes this label when it notices
   that the failed checks are no longer fresh/applicable.
-* S-merge-ready: Duplicates GitHub "green check" mark for the staging
-  branch commit. The bot removes this label before it updates the PR
-  target branch. Applied only if `config::skip_merge` option is on
-  (see below).
-* S-merge-failed: Unrecoverable failure other than the staging branch
-  test failure (the latter is marked with `S-staging-checks-failed`). The
-  bot ignores this PR until the PR branch or its target branch change.
+* `M-failed-description`: The PR title and/or description is invalid
+  (see below for PR commit message rules). The bot removes this label
+  when it revisits the PR and notices that the commit message components
+  were fixed.
+* `M-failed-other`: A fatal PR-specific error other than the staging
+  branch test failure (the latter is marked with
+  `M-failed-staging-checks`). (XXX: finishProcessing() sets this label
+  even if the staging branch test fail). It is probably necessary to
```

Comment:
I did not understand this XXX: finishProcessing() should be the only place where we know about failed staging checks. Probably another method was meant here, which was fixed at 19f4956.

Reply:
Thank you for fixing this.

Rephrasing XXX FYI: finishProcessing() sets `M-failed-other` when the staging branch test fail. It should not do that.